python/lsst/daf/butler/_butler.py

```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+        self,
+        refs: Iterable[DatasetRef],
+        /,
+        *,
+        full_check: bool = True,
+    ) -> dict[DatasetRef, DatasetExistence]:
+        """Indicate whether multiple datasets are known to Butler registry and
+        datastore.
+
+        This is an experimental API that may change at any moment.
+
+        Parameters
+        ----------
+        refs : iterable of `DatasetRef`
+            The datasets to be checked.
+        full_check : `bool`, optional
+            If `True`, an additional check will be made for dataset artifact
+            existence. This will involve additional overhead due to the need
+            to query an external system. If `False` registry and datastore
+            will solely be asked if they know about the dataset but no
+            check for the artifact will be performed.
+
+        Returns
+        -------
+        existence : dict of [`DatasetRef`, `DatasetExistence`]
+            Mapping from the given dataset refs to an enum indicating the
+            status of the dataset in registry and datastore.
+            Each value evaluates to `True` if the dataset is present and known
+            to both.
+        """
+        existence = {ref: DatasetExistence.UNRECOGNIZED for ref in refs}
+
+        # Registry does not have a bulk API to check for a ref.
+        for ref in refs:
+            registry_ref = self.registry.getDataset(ref.id)
+            if registry_ref is not None:
+                # It is possible, albeit unlikely, that the given ref does
+                # not match the one in registry even though the UUID matches.
+                # When checking a single ref we raise, but it's impolite to
+                # do that when potentially hundreds of refs are being checked.
+                # We could change the API to only accept UUIDs and that would
+                # remove the ability to even check and remove the worry
+                # about differing storage classes. Given the ongoing discussion
+                # on refs vs UUIDs and whether to raise or have a new
+                # private flag, treat this as a private API for now.
+                existence[ref] |= DatasetExistence.RECORDED
+
+        # Ask datastore if it knows about these refs.
+        knows = self.datastore.knows_these(refs)
+        for ref, known in knows.items():
+            if known:
+                existence[ref] |= DatasetExistence.DATASTORE
+
+        if full_check:
+            mexists = self.datastore.mexists(refs)
+            for ref, exists in mexists.items():
+                if exists:
+                    existence[ref] |= DatasetExistence._ARTIFACT
+        else:
+            # Do not set this flag if nothing is known about the dataset.
+            for ref in existence.keys():
+                if existence[ref] != DatasetExistence.UNRECOGNIZED:
+                    existence[ref] |= DatasetExistence._ASSUMED
+
+        return existence
+
+    @deprecated(
```

Comment:
I think I will have to remove this message for the moment since we have not filed an RFC formally deprecating the interface.

Reply:
Probably makes sense to add it on an implementation ticket for that RFC.